### DISCUSSION TITLE
Ubuntu20.04: pamltsp, ldap (via sssd), pam_mount problems

### DISCUSSION THREAD

**Discussion by URZNI** [ltsp/ltsp#675] [Q&A]
Hello there,
as part of a project I have LTSP (version 21.01) set up on a Ubuntu 20.04 system with chroot client images that are 20.04 as well.
As part of the client setup, I need them to mount individual cifs-directories from a Samba DC on the home directory of individual users. The login-process authenticates against an OpenLDAP via SSSD and the mount is handled by pam_mount.

I now face several problems, some of them are minor, some major and a lot of them very....weird. If this is not the right place for this, then I humbly apologize:

- "pamltsp: could not mount via sshfs"

Whenever sssd is activated and handles the authentication against an OpenLDAP, I get the above mentioned error message "pamltsp: could not mount via sshfs"...only for it to work perfectly without issue. My question here is, is that normal/expected? 
(there has been a github post about it, that has been resolved and the fixes put into the code but I cannot find it atm)
Additionally to that, even if I enable NFS in the ltsp.conf, the error-message still appears.

- groups: cannot find name for group ID "Placeholder"

After my users have logged in, they get the message above. I tried to implement the solution commented [here](https://github.com/ltsp/ltsp/discussions/668) but that does not work. My question here is: does this only work in the new ltsp version (22.01) or does this apply to older versions as well, since the [patchnotes](https://github.com/ltsp/ltsp/issues/1) do not mention anything similar.

- locale settings

After I change the language and keyboard settings of ubuntu (via locale-gen, dpkg-reconfigure or export) the client boots with these changes, but looses the keyboard-settings after login. Is this a common problem and does anybody know, what causes them?

- nautilus cannot be opened

While the authentication and mount at first works fine, the file manager (GUI) freezes upon opening and stops the proccess subsequently. It then opens it later on (after 10+ Minutes) but many of the home folders (Documents, Downloads etc.) are not accessible.

If you look at it now within the terminal, the samba-share is mounted and you can access the /home-directory of the user and all files perfectly fine (also including read/write permissions and correct IDs from the LDAP).

My pam_mount.conf.xml configuration is as follows:

`volume fstype="cifs" server="placeholder" path="HOME/%(USER)" mountpoint="/home/%(USER)" options = "workgroup=placeholder"`

I also tried out several other options but none worked for me.

Is this a known problem? Does anybody have an idea how to fix this or is Ubuntu intentionally designed to not allow mounts on home-directories?

PS: I really appreciate the work put into LTSP and am really grateful for everything that has been done. I also have to note, that I am not a native english speaker, so I apologize in advance for weird phrasing/spelling.

**Comment by alkisg** [comment]
Hi @URZNI,

LTSP provides its internal implementation of "pam_mount". Since you're not using it, you should disable it by putting `PAM_AUTH_TYPE=0` under `[clients]` in ltsp.conf to avoid all the "cannot mount via ssshfs" messages. The "there's a warning but it works" part is expected, as when you're using LDAP, LTSP then registers itself as an "additional" PAM hook that just mounts /home after the authentication is done, not a "primary" PAM hook that may cancel the login.
If you enable NFS /home and it's indeed mounted, then the message shouldn't appear. In older versions of LTSP the autodetection might not work well and you might need SSHFS=0.

Regarding groups, since your users come from LDAP, their groups will also come from there. LTSP shouldn't be related to that message that you see. Is that group that you see indeed an LDAP group, or was it defined in the chroot?

Regarding locale, which DM are you using, GDM? Try to login to a client in vt2, then run `sudo nano /usr/share/xsession/*.desktop`, then replace the Exec line with `Exec=/usr/bin/xterm`, so that you end up logging into a plain xterm. There, type `setxkbmap -query; locale` and paste the output here. For example, if it works properly on a plain xterm but not on a full session, maybe you have global GNOME settings that override the Xorg settings.

There's an issue where gnome-keyring-daemon doesn't work well with SSHFS and LTSP disables it at boot when it doesn't see an NFS /home mount. Since you're using samba homes, try to set SSHFS=0 to avoid that. This might or might not be related to the delay when launching Nautilus, but it shouldn't be related to the Desktop etc folder access. It is possible that Samba might be to blame for the rest of the issues that you see. Unfortunately I don't know of anyone using Samba homes, with or without LTSP, so I don't know if it's supposed to work or not.

Don't worry about your English it's fine. If native speakers ever complain that we're not fluent enough, we'll invite them to use one of our own native languages instead. :D

**Comment by URZNI** [comment]
Hello again,
thank you very much for your help @alkisg. I tried out most of what you suggested but it did not work out (or I implemented it wrong...):

> LTSP provides its internal implementation of "pam_mount". Since you're not using it, you should disable it by putting PAM_AUTH_TYPE=0 under [clients] in ltsp.conf to avoid all the "cannot mount via ssshfs" messages.

I did just that and it works, now I have the problem that my login is in a sort of "initial login"-state in which I have to configure the timezone, username etc. But if I go through that, then log out and log in again with a LDAP-user, the message is not shown.

> Regarding groups, since your users come from LDAP, their groups will also come from there. LTSP shouldn't be related to that message that you see. Is that group that you see indeed an LDAP group, or was it defined in the chroot?

The groups come from the LDAP. I tried out to create a local group (both on the server and the client chroot) with the same ID but after booting the group vanishes (I tried out putting `PWMERGE_DGR="placeholder"` in the `[common]` section of ltsp.conf but it did not work).

> Regarding locale, which DM are you using, GDM? Try to login to a client in vt2, then run sudo nano /usr/share/xsession/*.desktop, then replace the Exec line with Exec=/usr/bin/xterm, so that you end up logging into a plain xterm. There, type setxkbmap -query; locale and paste the output here.

Could you please elaborate on that? I know I use GDM (since I did not change it and 20.04 should still use that) but what exactly do you mean with vt2?
Regardless of that, I tried out some things and the locale problem only occurs with LDAP-users. Unfortunately I cannot test the commands you posted since the LDAP-users do not have root privileges (which makes debugging very tedious, I am sorry...).

> There's an issue where gnome-keyring-daemon doesn't work well with SSHFS and LTSP disables it at boot when it doesn't see an NFS /home mount. Since you're using samba homes, try to set SSHFS=0 to avoid that. 

Unfortunately, the Samba issue was not resolved with writing SSSHFS=0 in the ltsp.conf. I tried putting it into `[server]`, `[client]` or `[common]` but neither of them worked (regardless of `PAM_AUTH_TYPE`).

**Comment by alkisg** [comment]
vt2 means to press Ctrl+Atl+F2 and to login to the terminal there.

See the last paragraph of https://ltsp.org/man/ltsp.conf/ for one way to get root access on the clients.
So when you want to debug things, and you're logged in with an LDAP user, you can type `su -` and enter the root password that you defined in ltsp.conf and get root access.

Regarding the LDAP groups, you don't need to configure anything in LTSP other than PAM_AUTH_TYPE=0. You don't need the PWMERGE variables. If your groups are wrong, it means you haven't configured the LDAP client part properly inside your chroot.

**Comment by URZNI** [comment]
I now tried it in both vt2 as well as a normal session and both times got the same output:

```
root@ltsp42:/usr/bin# setxkbmap -query; locale
LAN=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL="en_US.UTF-8"
```

Interestingly, in vt2 the keyboard layout was the correct one while in the normal session it was changed into the wrong keyboard layout.

Regarding my weird PAM_AUTH_TYPE problem, have you encountered this one before?

> "[...] now I have the problem that my login is in a sort of "initial login"-state in which I have to configure the timezone, username etc. But if I go through that, then log out and log in again with a LDAP-user, the message is not shown."

**Comment by alkisg** [comment]
Note that the last GNOME version I've used was on Ubuntu 12.04; after that it got too heavy and bloated and didn't care about old hardware, it relied on opengl, it didn't respect the well established xorg settings, nor the "broadcasting before login" use case etc, so we've switched to Ubuntu MATE and we haven't looked back.
So I don't have experience with GNOME anymore.

I *think* that the LANG issue means that you didn't configure GNOME inside your chroot correctly. I.e. the problem isn't related to LTSP but to your chroot. You'll need to contact the GNOME community on how to configure GNOME chroots. Or just to VMs instead of chroots, which are easier to configure.
Regarding the "initial setup", doesn't GNOME show that once for all new users? I don't understand what the problem is there. If you create files as one user, and then reboot, are the files still there? If so, your /home/username works, and it's a GNOME issue, again not related to LTSP...

**Comment by URZNI** [comment]
I understand and will try to contact the gnome community for advice, maybe something comes out of it.
Nonetheless thank you very much for all your advice, what you wrote did clarify some things for me.
I understand that this forum is meant for problems/questions specifically to LTSP, I was only wondering if you had encountered something similar in the past (the combination of my problems with LTSP).
